hw/ip/otbn/doc/fcov.md

```diff
@@ -0,0 +1,238 @@
+- Load from a misaligned address
+
+### SW
+
+- Store to a valid address, where `<grs1>` is above the top of memory and a negative `<offset>` brings the load address in range.
+- Store to a valid address, where `<grs1>` is negative and a positive `<offset>` brings the load address in range.
+- Store to address zero
+- Store to the top word of memory
+- Store to an invalid address (aligned but above the top of memory)
+- Store to a misaligned address
+
+### BEQ
+
+All points should be crossed with branch taken / branch not taken.
+
+- Branch forwards
+- Branch backwards
+- Branch to a misaligned address (offset not a multiple of 4)
+- Branch forwards to an invalid address, above the top of memory
+- Branch backwards to an invalid address (wrapping past zero)
+- Branch to current address.
+- Branch instruction at end of a loop.
+
+The "branch to current address" item is problematic if we want to take the branch.
+Probably we need some tests with short timeouts to handle this properly.
+
+### BNE
+
+As for `BEQ`.
+
+### JAL
+
+- Jump forwards
+- Jump backwards
+- Jump to a misaligned address (offset not a multiple of 4)
+- Jump forwards to an invalid address, above the top of memory
+- Jump backwards to an invalid address (wrapping past zero)
+- Jump to current address.
+- Jump when the current PC is the top word in IMEM.
+- Jump instruction at end of a loop.
+
+Note that the "jump to current address" item won't be a problem to test since it will quickly overflow the call stack.
+
+### JALR
+
+- Jump with a positive offset
+- Jump with a negative offset
+- Jump with a misaligned base address which `<offset>` aligns (each of the 3 possible misalignments).
+- Jump with a large base address which wraps to a valid address by adding a positive `<offset>`.
+- Jump with a base address just above top of IMEM but wih a negative `<offset>` to give a valid target.
+- Jump with a negative offset, wrapping to give an invalid target.
+- Jump with a positive offset, giving an invalid target above top of IMEM.
+- Jump to current address.
+- Jump when the current PC is the top word in IMEM.
+- Jump instruction at end of a loop.
+
+Note that the "jump to current address" item won't be a problem to test since it will quickly over- or underflow the call stack, provided `<grd>` and `<grs1>` aren't both `x1`.
+
+### CSRRS
+
+- Write with a non-zero `bits_to_set` to each valid CSR.
+
+### CSRRW
+
+- Write to every valid CSR with a `<grd>` other than `x0`.
+- Write to every valid CSR with `<grd>` equal to `x0`.
+
+### ECALL
+
+No special coverage points for this instruction.
+
+### LOOP
+
+- Loop with a zero iteration count (causing an error)
+- Loop with a count of `'1` (the maximal value)
+- Loop when the match address would be above the top of memory.
+- Loop when the loop stack is full, causing an overflow.
+- Loop at the end of a loop.
+- Duplicate loop match address, matching top of stack
+- Duplicate loop match address, further down stack
```

Comment:
Assuming match address is end address have we architecturally defined what's happening here and do we want to? I I don't think the current ISA docs don't explicitly say what should happen though it should be easy enough for the RTL and ISS to match their behaviour.

Reply:
I assume we always just compare with the top of stack (so the RTL and ISS should probably just match). We don't special-case these cases in the spec, and I don't think there's any need to. However, I've just looked at our docs and we never actually say what LOOP and LOOPI do! (Oops...) I'll file a follow-up bug to remind us to do so.